Split media track and text track how-to guides into per-platform guides (#777)
THEOplayer/documentation: theoplayer/how-to-guides/roku/text-tracks/change-visible-captions.mdx

---
sidebar_position: 1
sidebar_label: Change the visible captions
---

# How to dynamically change the visible captions on Roku

import Intro from '../../shared/text-tracks/_change-visible-captions-intro.mdx';

<Intro />

## Usage

In the following example we iterate through all available text tracks.
Once we find a track with the desired language, we set its mode to `showing`.
After that, we assign the `textTracks` object back to the THEOplayer attribute.
This is all you have to do to make the desired text track visible.
You can also change the visible text track during playback.

```brightscript

> function setCaptionsLanguage(language as String)

Reviewer: Brightscript functions should return a value. If a method doesn't return a value, it should be a `sub` instead, like
```brightscript
sub setCaptionsLanguage(language as String)
   ...
end sub
```
This should be updated throughout the rest of the docs as well.

  textTracks = m.player.textTracks
  for i = textTracks.count() - 1 to 0 step -1
    if textTracks[i].kind = "captions" and textTracks[i].language = language then
      if m.fullScreen then
        textTracks[i].mode = "showing"

>       else
>         textTracks[i].mode = "hidden"
>       end if
>     else
>       textTracks[i].mode = "disabled"
>     end if
>   end for

Reviewer: Setting to disabled will make it unavailable to turn on.
```suggestion
  for i = textTracks.count() - 1 to 0 step -1
    if m.fullScreen and textTracks[i].kind = "captions" and textTracks[i].language = language
      textTracks[i].mode = "showing"
    else
      textTracks[i].mode = "hidden"
    end if
  end for
```

>   'assignment of new roAssociativeArray is required because roku deep-copied roAssociativeArray through fields (pass-by-value)

Reviewer: Make this change wherever else in the Roku documentation that this comment occurs
```suggestion
  ' Assignment of new roAssociativeArray is required, because Roku field values are copies, not references.
```

>   'read more: https://developer.roku.com/en-gb/docs/developer-program/performance-guide/optimization-techniques.md#OptimizationTechniques-DataFlow

Reviewer: Make this change wherever else in the Roku documentation that this comment occurs
```suggestion
  ' Read more: https://developer.roku.com/en-gb/docs/developer-program/performance-guide/optimization-techniques.md#OptimizationTechniques-DataFlow
```
